opm/autodiff/VFPHelpers.hpp

```diff
@@ -0,0 +1,970 @@
+    }
+}
+
+
+
+
+
+
+/**
+ * Computes the gfr parameter according to the gfr_type_
+ * @return Production rate of oil, gas or liquid.
+ */
+template <typename T>
+static T getGFR(const T& aqua, const T& liquid, const T& vapour,
+                  const VFPProdTable::GFR_TYPE& type) {
+    switch(type) {
+        case VFPProdTable::GFR_GOR:
+            // Gas-oil ratio = gas / oil
+            return zeroIfNan(vapour / liquid);
+        case VFPProdTable::GFR_GLR:
+            // Gas-liquid ratio = gas / (oil + water)
+            return zeroIfNan(vapour / (liquid + aqua));
+        case VFPProdTable::GFR_OGR:
+            // Oil-gas ratio = oil / gas
+            return zeroIfNan(liquid / vapour);
+        case VFPProdTable::GFR_INVALID: //Intentional fall-through
+        default:
+            OPM_THROW(std::logic_error, "Invalid GFR_TYPE: '" << type << "'");
+    }
+}
+
+
+
+
+
+
+/**
+ * Helper struct for linear interpolation
+ */
+struct InterpData {
+    InterpData() : ind_{0, 0}, inv_dist_(0.0), factor_(0.0) {}
+    int ind_[2]; //[First element greater than or equal to value, Last element smaller than or equal to value]
+    double inv_dist_; // 1 / distance between the two end points of the segment. Used to calculate derivatives and uses 1.0 / 0.0 = 0.0 as a convention
+    double factor_; // Interpolation factor
+};
+
+
+
+
+
+
+/**
+ * Helper function to find indices etc. for linear interpolation
+ */
+inline InterpData findInterpData(const double& value, const std::vector<double>& values) {
+    InterpData retval;
+
+    //If we only have one value in our vector, return that
+    if (values.size() == 1) {
+        retval.ind_[0] = 0;
+        retval.ind_[1] = 0;
+        retval.inv_dist_ = 0.0;
+        retval.factor_ = 0.0;
+    }
+    // Else search in the vector
+    else {
+        //First element greater than or equal to value
+        //Start with the second element, so that floor_iter does not go out of range
+        //Don't access out-of-range, therefore values.end()-1
+        auto ceil_iter = std::lower_bound(values.begin()+1, values.end()-1, value);
+
+        //Find last element smaller than range
+        auto floor_iter = ceil_iter-1;
+
+        //Find the indices
+        retval.ind_[0] = floor_iter - values.begin();
+        retval.ind_[1] = ceil_iter - values.begin();
+
+        //Find interpolation ratio
+        double dist = (*ceil_iter - *floor_iter);
+        if (std::abs(dist) > 0.0) {
+            //Possible source for floating point error here if value and floor are large,
+            //but very close to each other
+            retval.inv_dist_ = 1.0 / dist;
+            retval.factor_ = (value-*floor_iter) * retval.inv_dist_;
+        }
+        else {
+            retval.inv_dist_ = 0.0;
+            retval.factor_ = 0.0;
+        }
+    }
+
+    return retval;
+}
+
+
+
+
+
+
+/**
+ * An "ADB-like" structure with a single value and a set of derivatives
+ */
+struct adb_like {
```

Comment:
This could be replaced with an actual AD class at some point (perhaps when opm-material's Evaluation makes its way into opm-common, should we decide that is a good idea), but in the meantime it must follow the naming rules, ClassesAreNamedWithCapitalCamelCase.